homeassistant/components/automation/__init__.py

```diff
@@ -215,6 +232,109 @@ async def reload_service_handler(service_call):
+        self._condition_trace = None
+        self._config = config
+        self._context = context
+        self._error = None
+        self._state = "running"
+        self._timestamp_finish = None
+        self._timestamp_start = dt_util.utcnow()
+        self._trigger = trigger
+        self._unique_id = unique_id
+        self._variables = None
+
+    def set_error(self, ex):
+        """Set error."""
+        self._error = ex
+
+    def set_variables(self, variables):
+        """Set variables."""
+        self._variables = variables
+
+    def set_condition_trace(self, condition_trace):
+        """Set condition trace."""
+        self._condition_trace = condition_trace
+
+    def finished(self):
+        """Set finish time."""
+        self._timestamp_finish = dt_util.utcnow()
+        self._state = "stopped"
+
+    def as_dict(self):
+        """Return dictionary version of this AutomationTrace."""
+
+        action_traces = {}
+        condition_traces = {}
+        for key, trace_list in self._action_trace.items():
+            action_trace_list = []
+            for action_trace in trace_list:
+                action_trace_list.append(action_trace.as_dict())
+            action_traces[key] = action_trace_list
+
+        if self._condition_trace:
+            for key, trace_list in self._condition_trace.items():
+                condition_trace_list = []
+                for condition_trace in trace_list:
+                    condition_trace_list.append(condition_trace.as_dict())
+                condition_traces[key] = condition_trace_list
+
+        result = {
+            "action_trace": action_traces,
+            "condition_trace": condition_traces,
+            "config": self._config,
+            "context": self._context,
+            "state": self._state,
+            "timestamp": {
+                "start": self._timestamp_start,
+                "finish": self._timestamp_finish,
+            },
+            "trigger": self._trigger,
+            "unique_id": self._unique_id,
+            "variables": self._variables,
+        }
+        if self._error is not None:
+            result["error"] = str(self._error)
+        return result
+
+
+@contextmanager
+def trace_automation(hass, unique_id, config, trigger, context):
+    """Trace action execution of automation with automation_id."""
+    action_trace_clear()
+    action_trace = action_trace_get()
+    automation_trace = AutomationTrace(
+        unique_id, config, trigger, context, action_trace
+    )
+
+    if unique_id:
+        traces = hass.data[DATA_AUTOMATION_TRACE].setdefault(
+            unique_id, deque([], STORED_TRACES)
+        )
+        traces.append(automation_trace)
+
+    try:
+        yield automation_trace
+    except Exception as ex:  # pylint: disable=broad-except
+        if unique_id:
+            automation_trace.set_error(ex)
+        raise ex
+    finally:
+        if unique_id:
+            automation_trace.finished()
```

Comment:
If we only finish when we have a unique ID… why do we even bother collecting a trace if we have no ID ?

Reply:
It will complicate the code a bit if we check everywhere if we're tracing or not. Let's solve this in a follow-up PR making it possible to opt-out of tracing.